### DISCUSSION TITLE
Add CNN support

### DISCUSSION THREAD

**Discussion by nidhaloff** [nidhaloff/igel#76] [Ideas]
@Prasanna28Devadiga Let's discuss how we will add CNN support. Anyone is free to join and share ideas. 

I want to keep it very simple. My suggestion would be just to write CNN in the algorithm field in order to use a CNN model. Here is an example to illustrate what I mean:

```
model:
     type: classification
     algorithm: CNN # here just CNN insteead of NeuralNetClassifier. Everyone will recognize this because sklearn supports no CNN models.
```

Example of yaml configuration for CNN:
```
model:
    type: classification
    algorithm: CNN
    arguments:
        conv_layers:
            -
                num_filters: 32
                filter_size: 3
                pool_size: 2
                pool_stride: 2
            -
                num_filters: 64
                filter_size: 3
                pool_size: 2
                pool_stride: 2
                dropout: 0.4
        fc_layers:
            -
                fc_size: 128
                dropout: 0.4
```

**Comment by GouravWadhwa** [comment]
Hi @nidhaloff, For using CNN we might need the information of whether to use a pre-trained model or not and number of layers required. Will this be provided by the user?

**Reply by nidhaloff** [reply]
@GouravWadhwa yes eventually we will need that, thanks. I would suggest either writing the pre-trained model name in the algorithm field (e.g` algorithm: AlexNet`) or adding a new field/parameter (e.g` use-pre-trained: AlexNet`) below the algorithm field. Would you like to work on this feature?

**Reply by GouravWadhwa** [reply]
Yes I am trying to work on this feature. Further I am thinking of using tensorflow/keras for implementing the CNN models.
Currently I will just focus on the Pre-Trained models on classification problems after that we can maybe focus on random user designed models, and on the regression problem too.
Please let me know if I should add something else

**Reply by nidhaloff** [reply]
@GouravWadhwa sounds good ;) I have nothing to add at the moment

**Comment by Prasanna28Devadiga** [comment]
@GouravWadhwa this sounds great. 
@nidhaloff I have started working on implementing user-defined models. The way I see it, besides just coding up a CNN , we also need to find a way to work with image datasets. Keras can be used to perform pre-processing on the image dataset once its parsed by the CLI. Do you have any ideas about we might want to go about giving image inputs to the CLI, considering right now it only takes text inputs( csv , excel etc.)

**Reply by nidhaloff** [reply]
@Prasanna28Devadiga yes the dataset can be in csv format. 
I have an idea from my past experience. The dataset can be in csv format such that the image column will contain the path to the actual images. 

This means the users will first need to convert their images (or the folder that contains the actual images) to a csv file. A simple csv file would have two columns (image and label). Each row would have a path to an image and a label of that image. This csv file will be passed as the dataset and igel should read the file and use the path of the images to find the actual one, read them, pre-process them... 

Once we have a working example, we can also write an automated script to convert images to a csv dataset. 
There are many good libraries for image processing. I had a good experience with [ pillow](https://pillow.readthedocs.io/en/stable/) and matplotlib.

Is the approach clear?

**Reply by Prasanna28Devadiga** [reply]
Yes the approach is clear.

**Reply by Prasanna28Devadiga** [reply]
Hey @nidhaloff I am done coding up a CNN based on a small version of the Xception network. This should suffice for the basic needs of most users. Have also made sure to take inputs in the approach you mentioned above (csv with paths and labels). Do you have any particular such csv datasets you want me to test  it on? Also i might need a little guidance in making sure the CLI can recognize "CNN" as a valid form of input and make sure the cli is able to work with the model file.
Should I Create a pull request with the model code? You could then tell me how to proceed / things to incorporate.

**Reply by nidhaloff** [reply]
@Prasanna28Devadiga Is there a specific reason why you chose the Xception network? Does that mean you did implement a pre-trained model instead of a general-purpose model or am I wrong? 

In order to test your code locally, you can use the mnist dataset. Here is a basic setup:

first, clone the mnist images somewhere on your machine

```
git clone https://github.com/myleott/mnist_png.git
cd mnist_png/
tar -xf mnist_png.tar.gz
cd mnist_png/
```

after that, you can use [this gist](https://gist.github.com/nidhaloff/6537bb4c76790af86f0567435c8618b8) to convert the raw images to a csv dataset. You should end up having separate `mnist_dataset_training.csv` and `mnist_dataset_testing.csv `datasets.

**Reply by Prasanna28Devadiga** [reply]
@nidhaloff this isnt a pre-trained model. In a pre-trained model we have the weights for all of the layers set. However that is not the case. Here i have just made a general purpose model whose architecture resembles that of the Xception network. So whenever the user starts training, the model would be learning from scratch.

**Reply by nidhaloff** [reply]
@Prasanna28Devadiga Ah ok I see, but that means that the architecture is pre-defined, right?

**Reply by Prasanna28Devadiga** [reply]
@nidhaloff  Yes, does that not work? The architecture should be pretty powerful in itself considering that its been used in the official keras documentation. We could just add a simple for loop to keep adding more blocks (layers) however i think it would be redundant. Let me know what you think

**Reply by nidhaloff** [reply]
@Prasanna28Devadiga Although that works, it will not be always what the user needs. The architecture needs to be customizable. 

The thing is, if you have a task that a neural network consisted of 3 layers can solve, then why would you need 70 layers (70 is a random number here but you get what I mean). I'm working with ML every day and I can tell you that I would not use such an architecture to solve a "simple" problem. Therefore, I would let these parameters for the user to decide. Why should we decide what is the best architecture? there is no "best" architecture anyway. 

We can keep your implementation and use it maybe as a "default" if the user didn't configure the parameter. 
Here is an example of My idea/vision for a CNN example (something like this can be configured in the .yaml file):

```

# model definition
model:
    type: classification
    algorithm: CNN
    arguments:
        conv_layers:  # convolutional layers
            -   # if you are not familiar with yaml, the hyphen basically means that the conv_layers is a list of items/layers
                num_filters: 32
                filter_size: 3
                pool_size: 2
            -
                num_filters: 64
                filter_size: 3
                pool_size: 2
                pool_stride: 2
                dropout: 0.4
        fc_layers:  # fully connected layer
            -
                fc_size: 250
                dropout: 0.4
       
```

everything under arguments will be passed to the keras model, which will be used in training. This is my suggestion. Let me know what you think ( this may look complicated, but it's not, we just have to parse the yaml configuration and pass the arguments to the model)

**Reply by Prasanna28Devadiga** [reply]
@nidhaloff I get your point, I'll start working on making the architecture customizable.

**Reply by Prasanna28Devadiga** [reply]
@nidhaloff I'm almost done writing the code that makes a compeletely customizable architecture.
To make a 3 layer Neural Net, all the user would need to do is 
pass an argument (Dense(64),Dense(32),Dense(10)) and so on. Just have to specify the type of layer and all the parameters that are accepted.
Right now it can accept all types of Dense, Dropout, Conv , Pooling, Normalization layers that are supported by Keras. 
Does this work?

**Reply by Prasanna28Devadiga** [reply]
If so all that's required is passing the inputs from the yaml file to the model file. 
Also do you want to like have a
 "mode :default/customizable" parameter such that the user can choose whether they wanna go with a default CNN implementation we have or do we want to allow the users to be making their own CNNs each time. We could use the implementation i made earlier.
On second thoughts though, I doubt having a "default" CNN makes sense considering each user has different needs. Might be a better option to just have the user build their own CNN.
Let me know your thoughts too :)

**Reply by nidhaloff** [reply]
@Prasanna28Devadiga that's awesome. I would suggest that you work on the feature step by step. It does not have to be perfect. If you have a good example, then you can make a PR, which I can review and then give you feedback. You can always improve your work afterwards ;)

I agree that we don't need a default network architecture. I think we should let users decide that. However, I do suggest that we used default arguments if the user didn't provide any. Just like in keras there are some default parameters, that users do not have to provide, we should also provide default values for some parameters ( I suggest we use exactly the default values that keras use

I didn't understand what you mean by  (Dense(64),Dense(32),Dense(10)), but like I said if you have a working example, you can make a PR and I will review that and we can discuss if you need to change something.

**Reply by Prasanna28Devadiga** [reply]
By Dense(64) and so i meant that all a user needs to do is mention the type of layer(s) with the parameters inside parentheses and they'll be good to go. The model and training part is working fine , however there seems to be a bug with the reading dataset from a csv part. I'll try to get that fixed by tomorrow. Should i still create a pull request with 1 broken function?

**Reply by nidhaloff** [reply]
@Prasanna28Devadiga alright, no you should do a PR when everything is fine. 

I'm not exactly sure why did you choose that approach for passing the layer type. I find it "Ok" but not very convenient.
 Did you checked my previous comment(s)? I posted an example for defining the architecture, which is like this:

```
# model definition
model:
    type: classification
    algorithm: CNN
    arguments:
        conv_layers:  # convolutional layers
            -   # if you are not familiar with yaml, the hyphen basically means that the conv_layers is a list of items/layers
                num_filters: 32
                filter_size: 3
                pool_size: 2
            -
                num_filters: 64
                filter_size: 3
                pool_size: 2
                pool_stride: 2
                dropout: 0.4
        fc_layers:  # fully connected layer or you can name it Dense_layers if you want
            -
                fc_size: 64   # Dense layer size
```

That means your example with Dense(64) could be 
```
        Dense_layers:
            -
                size: 64   # Dense layer size
```

I find that more clear than 
```
        arguments:
               Dense(64)
```
because where would the user enter other parameters, like other types of layers (e.g conv) or the activation function, padding, etc...

**Reply by Prasanna28Devadiga** [reply]
I'm sorry for not being very clear. I meant to say the function takes Dense(64) as an input. Cant we parse the yaml file in a way such that Dense(64) is being sent to the function?

**Reply by nidhaloff** [reply]
@Prasanna28Devadiga of course, it should be pretty straightforward to do.

**Reply by Prasanna28Devadiga** [reply]
Then that wont be a problem I think. I'll try to fix the dataset function and try to setup the yaml part too asap.

**Reply by nidhaloff** [reply]
@Prasanna28Devadiga which problem are you having with the dataset? If you followed the steps I post, then you should already have a dataset example to start with

**Reply by Prasanna28Devadiga** [reply]
I did follow those steps and was able to get a csv file . However keras'  [flow_from_dataframe](https://keras.io/api/preprocessing/image/#flowfromdataframe-method) method is unable to recognize the images. The error i'm getting is "invalid filename" which on looking around online, seems to have a common fix i.e. providing the absolute paths in the csv file. 
However i dont think that would be the case each time and hence am stuck here.

**Comment by GouravWadhwa** [comment]
Hi @nidhaloff, in my experience pre-trained models are much more powerful than the models trained from scratch because of which they are used in general. Further, I also support the idea of adding customisable models in this library because of which I think we can give the option of using pre-trained model in yaml file. I the user does not wants to use a pre-trained model then they have to specify the model parameters.

**Reply by nidhaloff** [reply]
@GouravWadhwa What do you mean by **much more powerful?** Can you provide which experience you had with this?

If you mean that pre-trained models always achieve better performance, then you are wrong. Sometimes adding accidental complexity will make your model worse. Furthermore, the network will need much more time and energy to train. So why would a user need that if he/she has a problem that a smaller network can solve? 

Moreover, if you are using a pre-trained model, then you would have no idea what is happening, why your model is performing badly? what is the issue? On the other side, building a network step by step and seeing the differences in metrics for each step may help you as a user to figure out what is the best architecture for your problem. 

If you are working on adding support for pre-trained models, then you would have pre-defined network architecture, so this would not be a problem for pre-trained models. However, we need to implement this for customizable models (this is what @Prasanna28Devadiga is working on)

**Reply by GouravWadhwa** [reply]
@nidhaloff I was just saying if we have 2 models with same parameters then the pre-trained model generally performs better(I have worked on cross domain adaption and you would find all the state of the art works uses pre-trained feature extractors as these models have performed very nicely on some of the toughest datasets.). 
Also I suggested that in your proposed yaml file we can add a pertained_model argument in the model parameter which gives the pertained model we want to use if any otherwise the user can opt for there own configured model too. I was just suggesting of giving the user an extra choice of using pre-trained too.

**Comment by Shrimadh** [comment]
Hey @nidhaloff, as @Prasanna28Devadiga is working on User-defined models using Tensorflow/Keras.
I was wondering if I could work on the problem of user-defined models using **Pytorch** and use the custom dataset feature in pytorch to load up the images for training once the csv file is constructed using the [gist](https://gist.github.com/nidhaloff/6537bb4c76790af86f0567435c8618b8) mentioned above.
